<a href="https://colab.research.google.com/github/OLatinwo/REU-DataScienceProgram/blob/main/DataWrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lecture 1: Wrangling
  In this lecture notebook, we will test the hypothesis that the CEOs of major companies are typically in their 40s or older. To do this, we will identify top companies, find their CEOS, extract their ages, and look at the distribution of their ages. 

In [1]:
#Lets start by installing some libraries that are useful for programming data

#For crawling pages
!pip install scrapy

#Optional, for parallel excecution
!pip install swifter

#For string similarity
!pip install py_stringsimjoin

# lxml to parse xml tree
!pip install lxml

!pip install pandas
!pip install numpy
!pip install matplotlib

     |████████████████████████████████| 256kB 5.3MB/s 
     |████████████████████████████████| 256kB 45.2MB/s 
     |████████████████████████████████| 71kB 7.2MB/s 
     |████████████████████████████████| 3.2MB 37.4MB/s 
     |████████████████████████████████| 61kB 7.3MB/s 
     |████████████████████████████████| 3.1MB 29.3MB/s 
     |████████████████████████████████| 3.2MB 31.6MB/s 
     |████████████████████████████████| 81kB 8.4MB/s 
  Created wheel for PyDispatcher: filename=PyDispatcher-2.0.5-cp37-none-any.whl size=11517 sha256=5ba4065462c5d4d85a192b2729d45dfca8de155c0481d461961e8b3362db39ca
  Stored in directory: /root/.cache/pip/wheels/88/99/96/cfef6665f9cb1522ee6757ae5955feedf2fe25f1737f91fa7f
  Created wheel for protego: filename=Protego-0.1.16-cp37-none-any.whl size=7783 sha256=a6e410638f585a0ba273010bf72750feb0e3cf5f0cb20ff832bf880b7c8e6e2b
  Stored in directory: /root/.cache/pip/wheels/51/01/d1/4a2286a976dccd025ba679acacfe37320540df0f2283ecab12
Successfully built PyDispatch

     |████████████████████████████████| 1.1MB 4.6MB/s 
     |████████████████████████████████| 665kB 13.9MB/s 
  Created wheel for py-stringsimjoin: filename=py_stringsimjoin-0.3.2-cp37-cp37m-linux_x86_64.whl size=3723654 sha256=c9de6c4863debdfef1d705f98a9a99a6103197175f09cc9c13110573c639c0b5
  Stored in directory: /root/.cache/pip/wheels/58/30/8e/6fa463100ff6e8595b131067ec43a118c607f621ac297e7dad
  Created wheel for py-stringmatching: filename=py_stringmatching-0.4.2-cp37-cp37m-linux_x86_64.whl size=2055277 sha256=7e4056f6dea7dde43f2f68cc204348b354aa0e519755a552bf663654ac4bca92
  Stored in directory: /root/.cache/pip/wheels/be/8c/5d/bcdaf17d0784738b73e856ebc4061cd3f2509753a83dd71058
Successfully built py-stringsimjoin py-stringmatching


In [2]:
# Here are some imports for simplicity
# For parsing dates and being able to compare
import datetime

# For fetching remote data
import urllib

#Pandas dataframes
import pandas as pd

#Numpy matrix and array operations
import numpy as np

#Sqlite is a simplistic database
import sqlite3

#Crawler for multiple web pages at once
import scrapy
from scrapy.crawler import CrawlerProcess

#Try to parallalize the comptation
import swifter

#Approximate String Matching
import py_stringsimjoin as ssj
import py_stringmatching as sm

#Data Visualization
import matplotlib

#import urllib and etree for download and parsing 
from lxml import etree

# Acquiring Data About Companies and CEOs

In [3]:
data = urllib.request.urlopen(\
      'https://gist.githubusercontent.com/jvilledieu/c3afe5bc21da28880a30/raw/a344034b82a11433ba6f149afa47e57567d4a18f/Companies.csv')

company_data_df = pd.read_csv(data)

In [4]:
#Persist it so an SQLite database, and read it back.  
conn = sqlite3.connect('local.db')

company_data_df.to_sql("companies", conn, if_exists="replace", index=False)

pd.read_sql_query('select * from companies', conn)

permalink  ... last_funding_at
0                                   /organization/waywire  ...      30/06/2012
1                         /organization/tv-communications  ...      23/09/2010
2                           /organization/rock-your-paper  ...      09/08/2012
3                          /organization/in-touch-network  ...      01/04/2011
4                                   /organization/n-plusn  ...      04/09/2014
...                                                   ...  ...             ...
47753                                 /organization/zzish  ...      24/03/2014
47754  /organization/zznode-science-and-technology-co-ltd  ...      01/04/2012
47755                           /organization/zzzzapp-com  ...      10/09/2014
47756                          /organization/a-list-games  ...      21/11/2011
47757                                     /organization/x  ...      04/04/2013

[47758 rows x 18 columns]

In [5]:
# Now lets read a HTML table containing information about CEOs
company_ceos_df = pd.read_html('https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/List_of_chief_executive_officers.html#List_of_CEOs')[0]
company_ceos_df

Company  ...     Updated
0                     Accenture  ...  2019-01-31
1            Aditya Birla Group  ...  2018-10-01
2                 Adobe Systems  ...  2018-10-01
3                        Agenus  ...  2018-10-01
4                        Airbus  ...  2017-11-14
..                          ...  ...         ...
180             Williams-Sonoma  ...  2017-11-11
181  Wipro Technologies Limited  ...  2017-11-11
182                         WWE  ...  2017-11-11
183                 Yum! Brands  ...  2017-11-11
184                Zillow Group  ...  2018-12-10

[185 rows x 6 columns]

In [6]:
crawl_list = []
for executive in company_ceos_df['Executive']:
  crawl_list.append('https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/' + executive.replace(' ', '_') + '.html')
crawl_list

['https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Julie_Sweet.html',
 'https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Kumar_Birla.html',
 'https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Shantanu_Narayen.html',
 'https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Garo_H._Armen.html',
 'https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Tom_Enders.html',
 'https://raw.githubusercontent.com/odpi/OpenDS4All/penn-processing-zgi/assets/data/scripts-autoload-data/data-wrangling/en.wikipedia.org/wiki/Daniel_Zhang.html',
 'https://raw.githubu

In [7]:
# use urllib to crawl all pages in crawl_list, and store the response of page
# in list_pages
pages = []

for url in crawl_list: 
  page = url.split("/") [-1] # extract the persons name at the end of the url 
  print('Looking at file %s' % page)
  # split the url into different parts, and convert the persons name into ascii code
  url_list = list(urllib.parse.urlsplit(url))
  url_list[2] = urllib.parse.quote(url_list[2])
  url_ascii = urllib.parse.urlunsplit(url_list)
  try:
    response = urllib.request.urlopen((url_ascii))
    #Save the page and url for later use
    pages.append(response)
  except urllib.error.URLError as e:
    print(e.reason)

Looking at file Julie_Sweet.html
Looking at file Kumar_Birla.html
Looking at file Shantanu_Narayen.html
Looking at file Garo_H._Armen.html
Looking at file Tom_Enders.html
Looking at file Daniel_Zhang.html
Looking at file Jeff_Bezos.html
Looking at file Lisa_Su.html
Looking at file Stephen_Squeri.html
Looking at file Doug_Parker.html
Looking at file Joseph_R._Swedish.html
Looking at file Tim_Cook.html
Looking at file Lakshmi_Niwas_Mittal.html
Looking at file Randall_L._Stephenson.html
Looking at file Charles_Woodburn.html
Looking at file Tapan_Singhel.html
Looking at file Carlos_Torres_Vila.html
Looking at file Brian_Moynihan.html
Looking at file Jes_Staley.html
Looking at file Warren_Buffett.html
Looking at file Hubert_Joly.html
Looking at file Sunil_Bharti_Mittal.html
Looking at file Stephen_A._Schwarzman.html
Looking at file Andrew_Mackenzie.html
Looking at file Harald_Krüger.html
Looking at file Charles_Scharf.html
Looking at file Dennis_Muilenburg.html
Looking at file Rich_Lesser.h

In [8]:
pages

Populate a table with the name, we page URL, and birth date of each CEO

In [9]:
# Use etree.HTML(...) on the HTML content of each page to get a DOM tree that
# can be processed via XPath to extract the bday information. Store the CEO name, 
# webpage, and the birthday (born) in exec_df

# We first check that the HTML content has a table type vcard
# and then extract the bday information. If there is no birthday, the datetime 
# values is Nat (not a type)

exec_df = pd.DataFrame(columns=['name','page','born'])

for page in pages: 

  tree = etree.HTML(page.read().decode("utf-8")) #creatinf a DOM tree of the pages
  url = page.geturl()
  bday = tree.xpath('//table[contains(@class,"vacard")]//tbody/tr/td/span[@class="bday"]/text()')
  if (len(bday)>0):
    exec_df = exec_df.append({'name': url[url.rfind('/')+1:-5], 'page': url, 
                              'born': datetime.datetime.strptime(bday[0], '%Y-%m-%d')}, ignore_index=True)
  else:
    exec_df = exec_df.append({'name': url[url.rfind('/')+1:-5], 'page':url
                             , 'born': None}, ignore_index=True)
    
exec_df

name  ...  born
0            Julie_Sweet  ...  None
1            Kumar_Birla  ...  None
2       Shantanu_Narayen  ...  None
3          Garo_H._Armen  ...  None
4             Tom_Enders  ...  None
..                   ...  ...   ...
177       Laura_J._Alber  ...  None
178  Abidali_Neemuchwala  ...  None
179        Vince_McMahon  ...  None
180           Greg_Creed  ...  None
181      Spencer_Rascoff  ...  None

[182 rows x 3 columns]

# Cleaning and Manipulating the Tables

In [10]:
# Let's look at the name and born data in exec_df. We do this via projection. 
# Notice also (e.g row 24) that special characters will cause issues 

exec_df[['name', 'born']]

name  born
0            Julie_Sweet  None
1            Kumar_Birla  None
2       Shantanu_Narayen  None
3          Garo_H._Armen  None
4             Tom_Enders  None
..                   ...   ...
177       Laura_J._Alber  None
178  Abidali_Neemuchwala  None
179        Vince_McMahon  None
180           Greg_Creed  None
181      Spencer_Rascoff  None

[182 rows x 2 columns]

In [11]:
#Clean the name by removing underscores, and store in a new field called clean_name

exec_df['clean_name'] = exec_df['name'].apply(lambda x: x.replace('_', ' '))

exec_df

name  ...           clean_name
0            Julie_Sweet  ...          Julie Sweet
1            Kumar_Birla  ...          Kumar Birla
2       Shantanu_Narayen  ...     Shantanu Narayen
3          Garo_H._Armen  ...        Garo H. Armen
4             Tom_Enders  ...           Tom Enders
..                   ...  ...                  ...
177       Laura_J._Alber  ...       Laura J. Alber
178  Abidali_Neemuchwala  ...  Abidali Neemuchwala
179        Vince_McMahon  ...        Vince McMahon
180           Greg_Creed  ...           Greg Creed
181      Spencer_Rascoff  ...      Spencer Rascoff

[182 rows x 4 columns]

In [12]:
# We can do the same via SQL. For example, we'll save the dataframe first, and then extract just the name and clean_name using SQL

exec_df.to_sql('temp_exec', conn, if_exists="replace")

pd.read_sql_query('select name, replace(name, "_", " ") as clean_name from temp_exec', conn)

name           clean_name
0            Julie_Sweet          Julie Sweet
1            Kumar_Birla          Kumar Birla
2       Shantanu_Narayen     Shantanu Narayen
3          Garo_H._Armen        Garo H. Armen
4             Tom_Enders           Tom Enders
..                   ...                  ...
177       Laura_J._Alber       Laura J. Alber
178  Abidali_Neemuchwala  Abidali Neemuchwala
179        Vince_McMahon        Vince McMahon
180           Greg_Creed           Greg Creed
181      Spencer_Rascoff      Spencer Rascoff

[182 rows x 2 columns]

In [13]:
exec_df['clean_name']

0              Julie Sweet
1              Kumar Birla
2         Shantanu Narayen
3            Garo H. Armen
4               Tom Enders
              ...         
177         Laura J. Alber
178    Abidali Neemuchwala
179          Vince McMahon
180             Greg Creed
181        Spencer Rascoff
Name: clean_name, Length: 182, dtype: object

In [14]:
exec_df['clean_name']=='Kumar Birla'

0      False
1       True
2      False
3      False
4      False
       ...  
177    False
178    False
179    False
180    False
181    False
Name: clean_name, Length: 182, dtype: bool

In [15]:
exec_df[exec_df['clean_name']=='Kumar Birla']

name  ...   clean_name
1  Kumar_Birla  ...  Kumar Birla

[1 rows x 4 columns]

In [16]:
exec_df.to_sql('temp_exec', conn, if_exists='replace')
pd.read_sql_query('select * from temp_exec where clean_name="Kumar Birla"', conn)

index         name  ...  born   clean_name
0      1  Kumar_Birla  ...  None  Kumar Birla

[1 rows x 5 columns]

4. Putting it together
We start with a simple join between company_ceos_df and exec_df and persist it to the database. we check how many companies did not have a math CEO name. 

In [27]:
company_ceos_df[["Executive", "Company"]].merge(exec_df[["clean_name", "born"]],
                                                left_on=["Executive"],
                                                right_on=["clean_name"])

Executive                     Company           clean_name  born
0            Julie Sweet                   Accenture          Julie Sweet  None
1            Kumar Birla          Aditya Birla Group          Kumar Birla  None
2       Shantanu Narayen               Adobe Systems     Shantanu Narayen  None
3          Garo H. Armen                      Agenus        Garo H. Armen  None
4             Tom Enders                      Airbus           Tom Enders  None
..                   ...                         ...                  ...   ...
172       Laura J. Alber             Williams-Sonoma       Laura J. Alber  None
173  Abidali Neemuchwala  Wipro Technologies Limited  Abidali Neemuchwala  None
174        Vince McMahon                         WWE        Vince McMahon  None
175           Greg Creed                 Yum! Brands           Greg Creed  None
176      Spencer Rascoff                Zillow Group      Spencer Rascoff  None

[177 rows x 4 columns]

In [28]:
merged_df = company_ceos_df[["Executive", "Company"]].merge(exec_df[["clean_name", "born"]], 
                                                            left_on=["Executive"],
                                                            right_on=["clean_name"])
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
  print(merged_df['clean_name'])

0                         Julie Sweet
1                         Kumar Birla
2                    Shantanu Narayen
3                       Garo H. Armen
4                          Tom Enders
5                        Daniel Zhang
6                          Jeff Bezos
7                             Lisa Su
8                      Stephen Squeri
9                         Doug Parker
10                  Joseph R. Swedish
11                           Tim Cook
12               Lakshmi Niwas Mittal
13              Randall L. Stephenson
14                   Charles Woodburn
15                      Tapan Singhel
16                 Carlos Torres Vila
17                     Brian Moynihan
18                         Jes Staley
19                     Warren Buffett
20                        Hubert Joly
21                Sunil Bharti Mittal
22              Stephen A. Schwarzman
23                   Andrew Mackenzie
24                     Charles Scharf
25                  Dennis Muilenburg
26          

In [31]:
company_ceos_df.to_sql('company_ceos', conn, if_exists="replace")
exec_df.to_sql('executives', conn, if_exists="replace")

pd.read_sql_query('select Executive, Company, born from company_ceos ' +\
                  'join executives on Executive=clean_name', conn)

Executive                     Company  born
0            Julie Sweet                   Accenture  None
1            Kumar Birla          Aditya Birla Group  None
2       Shantanu Narayen               Adobe Systems  None
3          Garo H. Armen                      Agenus  None
4             Tom Enders                      Airbus  None
..                   ...                         ...   ...
172       Laura J. Alber             Williams-Sonoma  None
173  Abidali Neemuchwala  Wipro Technologies Limited  None
174        Vince McMahon                         WWE  None
175           Greg Creed                 Yum! Brands  None
176      Spencer Rascoff                Zillow Group  None

[177 rows x 3 columns]

Note: that the join above resulting in 177 rows. However, there are more rows in company_ceos_df so we are missing some companies. We can see which are missed when using a left outer join

In [37]:
leftjoin_df = company_ceos_df[["Executive", "Company"]].merge(exec_df[["clean_name", "born"]],
                                              left_on=["Executive"],
                                              right_on=["clean_name"], how="left", indicator=True)
leftjoin_df

Executive                     Company  ...  born _merge
0            Julie Sweet                   Accenture  ...  None   both
1            Kumar Birla          Aditya Birla Group  ...  None   both
2       Shantanu Narayen               Adobe Systems  ...  None   both
3          Garo H. Armen                      Agenus  ...  None   both
4             Tom Enders                      Airbus  ...  None   both
..                   ...                         ...  ...   ...    ...
180       Laura J. Alber             Williams-Sonoma  ...  None   both
181  Abidali Neemuchwala  Wipro Technologies Limited  ...  None   both
182        Vince McMahon                         WWE  ...  None   both
183           Greg Creed                 Yum! Brands  ...  None   both
184      Spencer Rascoff                Zillow Group  ...  None   both

[185 rows x 5 columns]

In [41]:
#Find names with invalid characters using with validation rules

replace_char = "Q"

failed = False
for name in exec_df['clean_name']:
  if not name.replace(' ', replace_char).\
          replace('.', replace_char).\
          replace('\'', replace_char).\
          replace('-', replace_char).isalpha():
    print("Illegal name %s"%name)
    failed = True

if failed:
    raise AssertionError('Found illegal names!')

Illegal name Harald Kr%C3%BCger
Illegal name Francisco D%27Souza
Illegal name B%C3%B6rje Ekholm
Illegal name Michael O%27Leary
Illegal name Matthias M%C3%BCller


AssertionError: ignored